### PR TITLE
Separate person from user

### LINKED ISSUE #55: Seperate Person from User
As a ÖA/Secretary I want to be able to add Persons to the database, so that people can find them even if those persons do not own an account on this website. I also want to be able to edit and delete the details of these persons later on.

### Acceptance
- [ ] Ability to create new POI on website
- [ ] Ability to edit/delete POI on website

### DoD
- [ ] Tests
- [ ] Documentation
- [ ] CI Build passing
- [ ] Peer Review
- [ ] Approval from PO

### PR DESCRIPTION
**Pull request by bewee**
Closes #55

### CODE REVIEW

**Review by Olliwehr — APPROVED** [review]
From our RH-Team perspective, our previous user-views are still working as intended now under the person/people-model. Good work 👍

File: app/models/user.rb [hunk 1]
```diff
@@ -11,12 +12,13 @@ class User < ApplicationRecord
...
-  has_one_attached :profile_picture
-  has_and_belongs_to_many :rooms
+  def initialize(attributes = {})
+    super(attributes)
+    return unless person.nil?
 
-  accepts_nested_attributes_for :openingtimes, allow_destroy: true
+    self.person = Person.new
+    person.email = email
+  end
```
Comment (taminob) [inline app/models/user.rb:21]: ugly, but as long as it works :)

File: db/migrate/20211209153121_create_join_table_people_rooms.rb [hunk 2]
```diff
@@ -0,0 +1,8 @@
+class CreateJoinTablePeopleRooms < ActiveRecord::Migration[6.1]
+  def change
+    create_join_table :people, :rooms do |t|
+      # t.index [:person_id, :room_id]
+      # t.index [:room_id, :person_id]
```
Comment (taminob) [inline db/migrate/20211209153121_create_join_table_people_rooms.rb:5]: why is this loop empty? i think these lines need to be added?

File: db/migrate/20211209165607_add_person_to_user.rb [hunk 3]
```diff
@@ -0,0 +1,5 @@
+class AddPersonToUser < ActiveRecord::Migration[6.1]
+  def change
+    add_reference :users, :person, null: true, foreign_key: true
```
Comment (taminob) [inline db/migrate/20211209165607_add_person_to_user.rb:3]: actually necessary?

File: spec/factories/people.rb [hunk 4]
```diff
@@ -0,0 +1,8 @@
+FactoryBot.define do
+  factory :person do
+    first_name { 'Herbert' }
+    last_name { 'Herbertson' }
+    phone_number { '+4990909090' }
+    email { "herbert.herbertson@hpi.de" }
```
Comment (taminob) [inline spec/factories/people.rb:6]: users didn't allow multiple users with the same mail, but people do now. Expected behavior or missing validation?

**Review by taminob — CHANGES_REQUESTED** [review]

**Review by joh-dah — APPROVED** [review]

File: app/models/person.rb [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+  has_and_belongs_to_many :rooms
+
+  accepts_nested_attributes_for :openingtimes, allow_destroy: true
+
+  def formatted_phone_number
+    parsed_phone = Phonelib.parse(phone_number)
+    return phone_number if parsed_phone.invalid?
+
+    parsed_phone.full_international
+  end
```
[the remark this answers is not in the document]
Comment (bewee) [reply app/models/person.rb:17]: Yeah... It wasn't covered before this PR and probably does not make so much sense as that test will only really test the external library, but for the sake of getting a green CI we can do it

File: app/models/user.rb (hunk 1 above)
Comment (bewee) [reply app/models/user.rb:21]: Ruby is weird^^

File: db/migrate/20211209153121_create_join_table_people_rooms.rb (hunk 2 above)
Comment (bewee) [reply db/migrate/20211209153121_create_join_table_people_rooms.rb:5]: I don't know, this code is auto-generated

File: db/migrate/20211209165607_add_person_to_user.rb (hunk 3 above)
Comment (bewee) [reply db/migrate/20211209165607_add_person_to_user.rb:3]: What exactly do you mean? The `null: true` part?

File: spec/factories/people.rb (hunk 4 above)
Comment (bewee) [reply spec/factories/people.rb:6]: Expected behavior. Multiple admins may want to register using their personal email addresses, but only provide admin@hpi.de on their public person.

File: app/controllers/people_controller.rb [hunk 6]
```diff
@@ -0,0 +1,69 @@
...
+
+  # Use callbacks to share common setup or constraints between actions.
+  def set_person
+    @person = Person.find(params[:id])
+  end
+
+  # Only allow a list of trusted parameters through.
+  def person_params
+    params.require(:person).permit(:phone_number, :first_name, :last_name, :email, :rooms, :profile_picture,
+                                   { openingtimes_attributes: [:id, :day, :opens, :closes] })
```
[the remark this answers is not in the document]
Comment (cambryx) [reply app/controllers/people_controller.rb:67]: In the form to edit people, we also want to edit the associated openingtimes. This is why we have to accept the attributes for openingtimes as well.

**Review by taminob — APPROVED** [review]

**Pull request merged**